classes/Api/ThankYous.php

```diff
@@ -0,0 +1,809 @@
+	 * @return bool
+	 */
+	public function CanSeeUser(SecurityContext $context, User $user): bool
+	{
+		return $this->acl->CanSeeUser($context, $user);
+	}
+
+	/**
+	 * @param SecurityContext $security_context
+	 * @return bool
+	 */
+	public function IsAdmin(SecurityContext $security_context): bool
+	{
+		return $this->acl->IsAdmin($security_context);
+	}
+
+	/**
+	 * Determines whether an Extranet Area is visible. If the second parameter is provided, this will be relative to
+	 * that Extranet.
+	 *
+	 * @param int      $target_extranet_id
+	 * @param int|null $viewers_extranet_id
+	 * @return bool
+	 */
+	public function IsExtranetVisible(int $target_extranet_id, ?int $viewers_extranet_id = null): bool
+	{
+		return $this->acl->IsExtranetVisible($target_extranet_id, $viewers_extranet_id);
+	}
+
+	/**
+	 * @param ThankYou $thank_you
+	 */
+	public function PopulateThankYouUsersFromThankables(ThankYou $thank_you)
+	{
+		$thankables = $thank_you->GetThankable();
+
+		if (!isset($thankables))
+		{
+			$thank_you->SetUsers(null);
+
+			return;
+		}
+
+		$owner_classes = [];
+		foreach ($thankables as $thankable)
+		{
+			$id        = $thankable->GetId();
+			$oclass_id = $thankable->GetOwnerClass();
+
+			if (!isset($id) || !isset($oclass_id))
+			{
+				continue;
+			}
+			$owner_classes[] = ['oclass' => $oclass_id, 'id' => $id];
+		}
+
+		$user_ids = $this->GetDistinctUserIdsFromOwnerClasses($owner_classes);
+
+		$users_list_provider = new UsersListProvider();
+		$users_list_provider->SetFilterIds($user_ids);
+
+		try
+		{
+			$users = $users_list_provider->GetListObjects();
+		} catch (InvalidFieldIsNotSingle $invalid_field_is_not_single)
+		{
+			throw new LogicException("Unexpected InvalidFieldIsNotSingle Exception throw by UserListProvider, GetListObjects", null, $invalid_field_is_not_single);
+		}
+
+		$thank_you->SetUsers($users);
+	}
+
+	/**
+	 * @param ThankYou $thank_you
+	 * @throws ThankYouException
+	 */
+	public function Notify(ThankYou $thank_you)
+	{
+		$thanked_users = $thank_you->GetUsers();
+		if (!isset($thanked_users))
+		{
+			throw new ThankYouException("Failed to Notify Thanked Users, Thanked Users haven't been set");
+		}
+
+		$all_users_ids = [];
+		foreach ($thanked_users as $thanked_user)
+		{
+			$all_users_ids[] = $thanked_user->GetId();
+		}
+
+		$description = $thank_you->GetDescription();
+
+		try
+		{
+			NotificationMessage::AddApplicationPrefix(Plugin::APPLICATION_NAME, Plugin::APPLICATION_NAME);
+
+			$params = [
+				'author'              => $thank_you->GetAuthor()->GetFullName(),
+				'other_people_number' => count($all_users_ids) - 1,
+				'description'         => $description
+			];
+			NotificationMessage::Send('thankyou.new_thanks', $params, $all_users_ids, self::IM_TYPE_THANKYOU);
+
+			if ($this->config->Get('notify_line_manager'))
+			{
+				$thankables = $thank_you->GetThankable();
+				if (!isset($thankables))
+				{
+					throw new ThankYouException("Failed to Notify Thanked User's Line Managers");
```

Comment:
Not sure why an exception needs to be thrown here and potentially interrupt the whole process of creating a thank you note, can this not just be logged as an error?

Reply:
The controller which calls this method currently catches said exception and logs it there. I considered changing the logging to be within this method, but I think the throwing of exceptions is clearer that the method has failed to perform its intentions.

I could swap the exceptions fired by this method to be `InvalidArgumentExceptions`, meaning that it's handled more like a typehinting error in the event that a ThankYou is passed in that doesn't have it's Thanked Users set, what d'you think :)?

Reply:
I'd argue that the `Notify` method should be protected, and not part of the public API. The controller shouldn't care about sending this, but the API layer should.